jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/PathMappings.java

```diff
@@ -123,29 +134,97 @@ public List<MappedResource<E>> getMatches(String path)
     {
         boolean isRootPath = "/".equals(path);
 
-        List<MappedResource<E>> ret = new ArrayList<>();
+        if (_mappings.isEmpty())
+            return Collections.emptyList();
+
+        List<MappedResource<E>> matches = null;
         for (MappedResource<E> mr : _mappings)
         {
             switch (mr.getPathSpec().getGroup())
             {
                 case ROOT:
                     if (isRootPath)
-                        ret.add(mr);
+                    {
+                        if (matches == null)
+                            matches = new ArrayList<>();
+                        matches.add(mr);
+                    }
                     break;
                 case DEFAULT:
                     if (isRootPath || mr.getPathSpec().matched(path) != null)
-                        ret.add(mr);
+                    {
+                        if (matches == null)
+                            matches = new ArrayList<>();
+                        matches.add(mr);
+                    }
                     break;
                 default:
                     if (mr.getPathSpec().matched(path) != null)
-                        ret.add(mr);
+                    {
+                        if (matches == null)
+                            matches = new ArrayList<>();
+                        matches.add(mr);
+                    }
                     break;
             }
         }
-        return ret;
+        return matches == null ? Collections.emptyList() : matches;
     }
 
     public MatchedResource<E> getMatched(String path)
+    {
+        if (_mappings.isEmpty())
+            return null;
+
+        if (_nonServletPathSpecs)
+            return getMatchedMixed(path);
+
+        if (_servletRoot != null && "/".equals(path))
+            return new MatchedResource<>(_servletRoot.getResource(), _servletRoot.getPathSpec(), _servletRoot.getPathSpec().matched(path));
+
+        MappedResource<E> candidate = _exactMap.getBest(path);
+        if (candidate != null)
+        {
+            MatchedPath matchedPath = candidate.getPathSpec().matched(path);
+            if (matchedPath != null)
+                return new MatchedResource<>(candidate.getResource(), candidate.getPathSpec(), matchedPath);
+        }
+
+        candidate = _prefixMap.getBest(path);
+        if (candidate != null)
+        {
+            MatchedPath matchedPath = candidate.getPathSpec().matched(path);
+            if (matchedPath != null)
+                return new MatchedResource<>(candidate.getResource(), candidate.getPathSpec(), matchedPath);
+        }
+
+        if (!_suffixMap.isEmpty())
+        {
+            int i = 0;
+            // Loop through each suffix mark
+            // Input is "/a.b.c.foo"
```

Comment:
What if the input is `/a.b.c/d.e.foo` ?

We don't want to do `path.indexOf('.', i+1)` within the path portion, only the last segment.
This needs test cases.

Reply:
Good question.    The impl for suffixes has not been changed, so this may be an existing problem.... well it is not really a problem, as I think it will give the right result, but will be inefficient as it will look for suffixes that could never exist.
I'll write another test case

Reply:
@joakime more tests added plus a small optimization for the "/a.b/c.d.foo" case.